### ISSUE TITLE
How to implement tensorflow/pytorch like convolution?

### ISSUE THREAD

**Issue by npuichigo** [arrayfire/arrayfire#2625]
I found that af::convolve only support the following modes of convolution. So how can we implement a convolution layer with kernel size (m, n, p, q)?
![image](https://user-images.githubusercontent.com/11533479/63739762-0b93f500-c8c1-11e9-9dfa-70a98fb81752.png)

For example, in pytorch's implementation of Conv2d, the input size is (channel_in, height, width) and the kernel size is (channel_in, channel_out, kernel_size_height, kernel_size_width), then we can get output with size (channel_out, height, width).
https://pytorch.org/docs/master/nn.html#torch.nn.Conv2d

**Comment by npuichigo** [comment]
Thank you for the quick response! Looking forward to the new changes.

**Comment by npuichigo** [comment]
By the way, for the following case, how to correspond to deep learning case like torch.nn.Conv2d? 

| Input Signal Dimensions   | Filter Dimensions   | Output Dimensions   | Batch Mode                | Description                                               |
| :-----------------------: | :-----------------: | :-----------------: | :-----------------------: | :-------------------------------------------------------- |
| [m n p qs]                | [m n p qf]          | [m n qs qf]         | Batch qs * qf             | qs signals and qf filsters to create qs * qf results      |

Does that means if we have input signal with shape [H, W, C_in, 1] and kernel with shape [K_h, K_w, C_in, C_out], then we can get output with shape [H, W, 1, C_out]->[H, W, C_out].

Or we have input with shape [H, W, 1, C_in] and the same kernel, and we get output with shape [H, W, C_in, C_out] instead. Then we need to sum over dimension 2 to get [H, W, 1, C_out]->[H, W, C_out].

**Comment by 9prady9** [comment]
@npuichigo We have updated documentation for convolve2 (All convolve* fns infact to reflect more accurate possible batching operations).

For convolve2, please look at the following link
https://github.com/arrayfire/arrayfire/blob/master/docs/details/signal.dox#L141

Those (tabled in the above file) are the only kind of batching operations possible.

Coming to the table entry you mentioned, it looks like a [3d convolution](https://github.com/arrayfire/arrayfire/blob/master/docs/details/signal.dox#L166) to me but this kind of disparate batch sizes(both qs and qf > 1) for 4th dimension are not currently supported in ArrayFire.

Having said that, the example you explained (shared below for quick reference)
```
Input - [ H W C_in 1]
Kernel - [K_h, K_w, C_in, C_out]
```
is basically a 3d convolve and it is batch operation in second row of the table mentioned here - https://github.com/arrayfire/arrayfire/blob/master/docs/details/signal.dox#L169

At this point ArrayFire will compute an output array of dimensions `[H W C_in C_out]`, now calling a 
`af::sum([H W C_in C_out], 2)` along dimension 2, will give what you are calling `[H, W, 1, C_out]`.
To reshape it, you need a no-op function call `moddims(out, dim4(H W C_out))`.

Hope that helps and please let me know if I misunderstand you explanation anywhere.

**Comment by 9prady9** [comment]
fyi - More generalized conv2 which handles disparate batch sizes is what @syurkevi referred to from #2359

**Comment by npuichigo** [comment]
@9prady9 @syurkevi I tried your suggestion to use 3d convolution, but after comparison, I found the result is not the same as PyTorch. (https://pytorch.org/docs/master/nn.html#torch.nn.Conv2d)

In this case, pytorch requires that the input tensor to be **[N, C_in, H_in, W_in] = [1, 3, 5, 4]** and the kernel to be **[C_in, C_out, K_h, K_w] = [3, 2, 3, 3]**. The output will be with shape **[N, C_out, H_in, W_in] = [1, 2, 5, 4]**. Just ignore the batch size of axis 0 because I set that to 1.

pytorch:
![image](https://user-images.githubusercontent.com/11533479/63747366-8b2dbe00-c8d9-11e9-97ee-ba08c6e84ec4.png)

The corresponding input and kernel in arrayfire is **[H_in, W_in, C_in, 1] = [5, 4, 3, 1]** and **[K_h, K_w, C_in, C_out] = [3, 3, 3, 2]**, so we get the output of **[H_in, W_in, C_in, C_out] = [5, 4, 3, 2]**. Then we do the sum to get **[H_in, W_in, 1, C_out] = [5, 4, 1, 2]** and reshape it to **[H_in, W_in, C_out ] = [5, 4, 2]**.

arrayfire:
![image](https://user-images.githubusercontent.com/11533479/63747555-01cabb80-c8da-11e9-8444-944d30b3c462.png)

![image](https://user-images.githubusercontent.com/11533479/63747584-0e4f1400-c8da-11e9-9b19-52d7b8797b12.png)

**Comment by 9prady9** [comment]
Torch convn is probably different afterall, in any case I will try to understand how its different from ArrayFire. #2359 Should address your concern I think.

**Comment by 9prady9** [comment]
@npuichigo 

The idea suggested actually works, I mistakenly suggested convolve3 instead of convolve2. Here is the sample output on my machine on v3.6.4 tag (output should match our installer binary computation).
```bash
tempO
[5 4 3 2]
    4.0000     6.0000     6.0000     4.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    4.0000     6.0000     6.0000     4.0000 

    4.0000     6.0000     6.0000     4.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    4.0000     6.0000     6.0000     4.0000 

    4.0000     6.0000     6.0000     4.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    4.0000     6.0000     6.0000     4.0000 


    4.0000     6.0000     6.0000     4.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    4.0000     6.0000     6.0000     4.0000 

    4.0000     6.0000     6.0000     4.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    4.0000     6.0000     6.0000     4.0000 

    4.0000     6.0000     6.0000     4.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    6.0000     9.0000     9.0000     6.0000 
    4.0000     6.0000     6.0000     4.0000 



output
[5 4 1 2]
   12.0000    18.0000    18.0000    12.0000 
   18.0000    27.0000    27.0000    18.0000 
   18.0000    27.0000    27.0000    18.0000 
   18.0000    27.0000    27.0000    18.0000 
   12.0000    18.0000    18.0000    12.0000 


   12.0000    18.0000    18.0000    12.0000 
   18.0000    27.0000    27.0000    18.0000 
   18.0000    27.0000    27.0000    18.0000 
   18.0000    27.0000    27.0000    18.0000 
   12.0000    18.0000    18.0000    12.0000 
```

**Comment by npuichigo** [comment]
@9prady9 Thanks! But if I use convolve2, which of the following is the corresponding situation?

| Input Signal Dimensions | Filter Dimensions | Output Dimensions | Batch Mode              | Description |
|:-----------------------:|:-----------------:|:-----------------:|:-----------------------:|:------------|
| [m n 1 1]               | [m n 1 1]         | [m n 1 1]         | No Batch                | Output will be a single convolved array |
| [m n 1 1]               | [m n p 1]         | [m n p 1]         | Filter is Batched       | p filters applied to same input |
| [m n p 1]               | [m n 1 1]         | [m n p 1]         | Signal is Batched       | 1 filter applied to p inputs |
| [m n p q]               | [m n p q]         | [m n p q]         | Identical Batches       | p*q filters applied to p*q inputs in one-to-one correspondence |
| [m n p 1]               | [m n 1 q]         | [m n p q]         | Non-overlapping batches | q filters applied to p inputs in to produce p x q results |
| [m n 1 p]               | [m n q 1]         | [m n q p]         | Non-overlapping batches | q filters applied to p inputs in to produce q x p results |

In fact, if I set the kernel to [K_h, K_w, 1, C_out], I can just get the same result (because we use all-one matrices as signal and kernel). Maybe the result can be different if we use different signal and kernel.

Actually, the difference is how convolution is computed with unequal input and output channels. In deep learning, we often use C_in number of filter to convolve the corresponding signal with C_in channels, followed with the sum operation.
![test](https://user-images.githubusercontent.com/11533479/63836740-b2a18b00-c9ac-11e9-8537-62b6aa2f8407.gif)

**Comment by 9prady9** [comment]
I think the table can be improved further, but to answer your question - it loosely falls under the category of last but one row where we do allow kernel of dimensions `[ a b c d]` - not clearly explained in the docs, hence my comment about further improvement.

I will send a PR soon to fix the documentation.

The answer didn't come out accidentally correct because of ones, the shapes we discussed
```
Input - [ H W C_in 1]
Kernel - [K_h, K_w, C_in, C_out]
```
do indeed work. I don't have torch setup, but you are welcome to try a random valued array for test and the values should be computed correctly.

For ArrayFire (channels i.e. slices in ArrayFire jargon) are independently computed - thus we don't expect anything like dimensions beyond 2 (since it is 2d convolution here) have to match. Hence, the `af::sum(..., 2)` call in the code I shared earlier. This sum is precisely the summation of results of respective channels that is show in the gif shared.
